File: sanguo/frameworks/runtime-src/Classes/Network/ark_socketEvent.cpp

#include "ark_socketEvent.h"


void ark_socketEvent::OnConnect(ClientSocket* s)
{
	LuaScriptData data(getNode(), "cc.Node");
	data.push("string", (void*)"OnConnect");
	data.push("string", (void*)s->GetMsg().c_str());
	ScriptEvent scriptEvent(kCustomLuaEvent, &data);
	ScriptEngineManager::getInstance()->getScriptEngine()->sendEvent(&scriptEvent);
}
void ark_socketEvent::OnConnectFail(ClientSocket* s)
{
	LuaScriptData data(getNode(), "cc.Node");
	data.push("string", (void*)"OnConnectFail");
	//data.push("string", (void*)s->GetMsg().c_str());
	ScriptEvent scriptEvent(kCustomLuaEvent, &data);
	ScriptEngineManager::getInstance()->getScriptEngine()->sendEvent(&scriptEvent);
}
void ark_socketEvent::OnSocketError(ClientSocket* s)
{
	LuaScriptData data(getNode(), "cc.Node");
	data.push("string", (void*)"OnSocketError");
	data.push("string", (void*)s->GetMsg().c_str());
	ScriptEvent scriptEvent(kCustomLuaEvent, &data);
	ScriptEngineManager::getInstance()->getScriptEngine()->sendEvent(&scriptEvent);
}
void ark_socketEvent::OnClose(ClientSocket* s)
{
	LuaScriptData data(this, "ark_socketEvent");
	data.push("string", (void*)"OnClose");
	data.push("string", (void*)s->GetMsg().c_str());
	ScriptEvent scriptEvent(kCustomLuaEvent, &data);
	ScriptEngineManager::getInstance()->getScriptEngine()->sendEvent(&scriptEvent);
}
//接收
void ark_socketEvent::OnData(ClientSocket* s, const char* data, int dataLen)
{
	ark_NetMsg* pMsg = new ark_NetMsg((unsigned char*)(data), dataLen);
	NetMsgMgr::getInstance()->addRecvMsg(pMsg);
}

//发送
void ark_socketEvent::OnRunExData(ClientSocket* s)
{
	ark_NetMsgPtr pMsg = NetMsgMgr::getInstance()->extractSendMsg();
	if (!pMsg) return;

	unsigned char* p = (unsigned char*)pMsg->getBuffer();
	int size = s->Send((const char*)p, pMsg->getSize());
	if (size != (int)pMsg->getSize())
		log("ClientSocket::send Error sentSize=%d, msgSize=%d", size, pMsg->getSize());
	delete pMsg;
}


int ark_socketEvent::getOnRecPack()
{
	tempStream = nullptr;
	lastpMsg = nullptr;
	NetMsgMgr* netMgr = NetMsgMgr::getInstance();
	if (!netMgr) return 0;
	ark_NetMsgPtr pMsg = netMgr->extractRecvMsg();
	if (!pMsg)
		return 0;
	//ark_Stream stream;
	tempStream = new ark_Stream();
	tempStream->CreateReadStreamFromBuf((const char*)pMsg->getBuffer(), pMsg->getSize());
	lastpMsg = pMsg;
	//这里不释放，releaseStream 里面释放tempStream时候ark_Stream的析构会释放m_stream（即pMsg的buf）
	//delete pMsg;  
	return 1;
}

File: sanguo/frameworks/cocos2d-x/cocos/scripting/lua-bindings/auto/api/ark_socketEvent.lua

--------------------------------
-- @module ark_socketEvent
-- @extend ClientSocketEvent
-- @parent_module 

--------------------------------
-- 
-- @function [parent=#ark_socketEvent] getOnRecPack 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#ark_socketEvent] OnRunExData 
-- @param self
-- @param #ClientSocket s
-- @return ark_socketEvent#ark_socketEvent self (return value: ark_socketEvent)
        
--------------------------------
-- 
-- @function [parent=#ark_socketEvent] OnConnect 
-- @param self
-- @param #ClientSocket s
-- @return ark_socketEvent#ark_socketEvent self (return value: ark_socketEvent)
        
--------------------------------
-- 
-- @function [parent=#ark_socketEvent] OnSocketError 
-- @param self
-- @param #ClientSocket s
-- @return ark_socketEvent#ark_socketEvent self (return value: ark_socketEvent)
        
--------------------------------
-- 
-- @function [parent=#ark_socketEvent] OnClose 
-- @param self
-- @param #ClientSocket s
-- @return ark_socketEvent#ark_socketEvent self (return value: ark_socketEvent)
        
--------------------------------
-- 
-- @function [parent=#ark_socketEvent] releaseStream 
-- @param self
-- @return ark_socketEvent#ark_socketEvent self (return value: ark_socketEvent)
        
--------------------------------
-- 
-- @function [parent=#ark_socketEvent] OnData 
-- @param self
-- @param #ClientSocket s
-- @param #char data
-- @param #int dataLen
-- @return ark_socketEvent#ark_socketEvent self (return value: ark_socketEvent)
        
--------------------------------
-- 
-- @function [parent=#ark_socketEvent] OnConnectFail 
-- @param self
-- @param #ClientSocket s
-- @return ark_socketEvent#ark_socketEvent self (return value: ark_socketEvent)
        
--------------------------------
-- 
-- @function [parent=#ark_socketEvent] getNode 
-- @param self
-- @return Node#Node ret (return value: cc.Node)
        
--------------------------------
-- 
-- @function [parent=#ark_socketEvent] getStream 
-- @param self
-- @return ark_Stream#ark_Stream ret (return value: ark_Stream)
        
--------------------------------
-- 
-- @function [parent=#ark_socketEvent] ark_socketEvent 
-- @param self
-- @return ark_socketEvent#ark_socketEvent self (return value: ark_socketEvent)
        
return nil
